### PR TITLE
Doc 29 exemple slurm job arrays

### PR DESCRIPTION
**Pull request by notoraptor**
Salut @lebrice ! Voici la PR pour DOC-29 avec les job arrays! https://mila-iqia.atlassian.net/browse/DOC-29

### CODE REVIEW

**Comment by notoraptor** [comment]
Et voici le rendu actuel: https://mila-docs--225.org.readthedocs.build/en/225/examples/good_practices/slurm_job_arrays/index.html

File: docs/examples/good_practices/slurm_job_arrays/main.py [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+from torchvision import transforms
+from torchvision.datasets import CIFAR10
+from torchvision.models import resnet18
+from tqdm import tqdm
+
+
+def main():
+    array_task_id = os.environ["SLURM_ARRAY_TASK_ID"]
+
+    training_epochs = int(array_task_id)
```
Comment (lebrice) [inline docs/examples/good_practices/slurm_job_arrays/main.py:20]: C'est pas exactement ça qu'on veut changer d'une job à l'autre. Je pense qu'il y aurait de meilleures options:

- changer des paramêtres en les selectionnant dans une liste à l'index donné
- changer le random seed pour les mêmes paramêtres
- encore mieux: changer le random seed qui serait utilisé pour sample les paramêtres aléatoirement.

Je ne me souviens pas si l'exemple de base ajoute le random seed, mais sinon ce serait bien de le faire ici.

**Review by lebrice — CHANGES_REQUESTED** [review]
Merci @notoraptor ! J'ai laissé un commentaire.

File: docs/examples/good_practices/slurm_job_arrays/job.sh [hunk 2]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+#SBATCH --gpus-per-task=rtx8000:1
+#SBATCH --cpus-per-task=4
+#SBATCH --ntasks-per-node=1
+#SBATCH --mem=16G
+#SBATCH --time=00:15:00
+
+# No need to change this file, as environment variables
```
Comment (lebrice) [inline docs/examples/good_practices/slurm_job_arrays/job.sh:8]: Est-ce que ça c'est dans le single-gpu example?

File: docs/examples/generate_diffs.sh [hunk 3]
```diff
@@ -50,4 +50,8 @@ generate_diff distributed/single_gpu/main.py good_practices/wandb_setup/main.py
 generate_diff distributed/single_gpu/job.sh frameworks/jax/job.sh
 generate_diff distributed/single_gpu/main.py frameworks/jax/main.py
 
+# single_gpu -> slurm job arrays
```
Comment (lebrice) [inline docs/examples/generate_diffs.sh:53]: Pense tu que ça aurait de l'allure de montrer cet exemple en diff par rapport à l'exemple de ton autre PR qui apprend comment passer les args au script avec `$@`? Ou single-gpu? (Aurais-tu une opinion là-dessus @obilaniu par hasard?)

File: docs/examples/good_practices/slurm_job_arrays/main.py [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+
+def main():
+    array_task_id = os.environ["SLURM_ARRAY_TASK_ID"]
+
+    training_epochs = 10
+    learning_rate = 5e-4
+    weight_decay = 1e-4
+    batch_size = 128
+
+    # Seed the random number generators as early as possible.
```
Comment (lebrice) [inline docs/examples/good_practices/slurm_job_arrays/main.py:27]: Je pense que ce serait mieux d'utiliser le job index pour faire le seed qui génère (seulement) les hyper-paramêtres, et voici pourquoi:
Après qu'on fasse ça, on pourra alors utiliser le **task id** (quand `--ntasks>1` pour obtenir le different seed (pour le reste de l'expérience) pour les mêmes paramêtres à un index dans le array.

Exemple: lancer 5 jobs, (donc 5 sets d'hyper-paramêtres différents) chacunes avec deux seeds (deux résultats différents avec les mêmes paramêtres).

On pourrait aussi expliquer dans le texte de l'exemple qu'ici on utilise les job arrays pour que chaque job (chaque index) obtienne des hyper-paramêtres différents, dans le contexte d'un sweep de HPO.

**Review by lebrice — CHANGES_REQUESTED** [review]

File: docs/examples/good_practices/slurm_job_arrays/job.sh (hunk 2 above)
Comment (notoraptor) [reply docs/examples/good_practices/slurm_job_arrays/job.sh:8]: Non ! J'ai rajouté ces commentaires pour préciser qu'il n'est pas nécessaire de modifier `job.sh` pour cet exemple !

Et aussi, quand le `job.sh` n'a aucune différence, j'ai des erreurs avec le `pre-commit`, il refuse de générer les diffs. Je ne sais pas encore pourquoi, du coup, j'ai ajouté ce commentaire pour contourner le problème (pour qu'il y ait au moins une différence entre les deux fichies).

File: docs/examples/generate_diffs.sh (hunk 3 above)
Comment (notoraptor) [reply docs/examples/generate_diffs.sh:53]: Oh, peut-être. Dans ce cas, je peux changer le titre de cette PR-ci en WIP, et attendre que l'autre soit reviewée puis mergée avant de rebaser celle-ci sur l'autre exemple.

File: docs/examples/good_practices/slurm_job_arrays/main.py (hunk 4 above)
Comment (notoraptor) [reply docs/examples/good_practices/slurm_job_arrays/main.py:27]: Ok ! Si je ne me trompe pas, ici, les hyper-paramètres sont ces valeurs-ci, n'est-ce pas ?

```python
    training_epochs = 10
    learning_rate = 5e-4
    weight_decay = 1e-4
    batch_size = 128
```

Donc, je dois utiliser le job index pour générer chacune d'elle aléatoirement ?

Comment (lebrice) [reply docs/examples/good_practices/slurm_job_arrays/main.py:27]: Ouais, tout ceux-cis sauf `training_epochs`. (En passant, juste pour le fun, celui-la serait un bon paramêtre de type "fidelity" pour Orion)

Comment (notoraptor) [reply docs/examples/good_practices/slurm_job_arrays/main.py]: Ok. J'ai fait un commit. Actuellement, ça donne ceci: https://mila-docs--225.org.readthedocs.build/en/225/examples/good_practices/slurm_job_arrays/index.html

Je ne sais pas s'il y a une manière plus officielle/académique de générer les valeurs ?

Comment (notoraptor) [reply docs/examples/good_practices/slurm_job_arrays/main.py]: @lebrice Nouveau commit. Qu'en penses-tu? https://github.com/mila-iqia/mila-docs/pull/225/commits/e8dc7b2039078de72143ed4eb5708e4b05ae9d28

File: docs/examples/good_practices/slurm_job_arrays/job.sh (hunk 2 above)
Comment (lebrice) [reply docs/examples/good_practices/slurm_job_arrays/job.sh:8]: Hmm si il n'y a aucune différence, alors pas besoin d'inclure ce fichier dans l'exemple (et donc pas besoin de créer de diff non plus)

File: docs/examples/good_practices/slurm_job_arrays/main.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+def main():
+    # Use SLURM ARRAY TASK ID to create a random number generator.
+    array_task_id = int(os.environ["SLURM_ARRAY_TASK_ID"])
+    gen = numpy.random.default_rng(seed=array_task_id)
+
+    training_epochs = 10
+    # Use random number generator to generate hyper-parameters.
+    learning_rate = gen.random() * 1e-3
+    weight_decay = gen.random() * 1e-3
```
Comment (lebrice) [inline docs/examples/good_practices/slurm_job_arrays/main.py:27]:
```suggestion
    learning_rate = gen.uniform(1e-6, 1e-2)
    weight_decay = gen.uniform(1e-6, 1e-3)
```

**Review by lebrice — APPROVED** [review]
Merci!

File: docs/examples/good_practices/slurm_job_arrays/main.py (hunk 5 above)
Comment (notoraptor) [reply docs/examples/good_practices/slurm_job_arrays/main.py]: Fait! https://github.com/mila-iqia/mila-docs/pull/225/commits/e1be4cadd7ac29a5f09e3d3fc2e3a5b3722edb5c

File: docs/examples/good_practices/slurm_job_arrays/job.sh (hunk 2 above)
Comment (notoraptor) [reply docs/examples/good_practices/slurm_job_arrays/job.sh]: Ok ! Fait ! https://github.com/mila-iqia/mila-docs/pull/225/commits/3e9a5ce922debe9d2400cf30f983711ca6651e0d

**Review by obilaniu — APPROVED** [review]

File: docs/examples/good_practices/slurm_job_arrays/main.py (hunk 1 above)
Comment (satyaog) [reply docs/examples/good_practices/slurm_job_arrays/main.py]: Small comment here. If a user needs the `job.sh` to run the example it should still be found in the directory?

File: docs/examples/good_practices/slurm_job_arrays/job.sh (hunk 2 above)
Comment (satyaog) [reply docs/examples/good_practices/slurm_job_arrays/job.sh]: Small comment here. If a user needs the `job.sh` to run the example it should still be found in the directory right?
